In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 5.3 MB/s 
     |████████████████████████████████| 49 kB 5.4 MB/s 


In [2]:
from pyomo.environ import *


In [3]:
import numpy as np

In [4]:
#create model 
model=ConcreteModel()

In [5]:
#number of dicision  variable
N=12
#number of constraint
M=5


In [6]:
# coefficient for objecetive function can provided using numpy arraysbe 
obj_coef_c=np.array( [0.56, 1.721, -9.6, -2.85, 41.5, 33.05, -0.75, 20.5, -0.35, -31.02, 50.65, 6.125])

In [7]:
#We use more numpy arrays for providing constraint coefficients
constr1_coef_A=np.array([[1, -1, 0, 0, 2, -2, 1, 0,0, -1, 1, -2],
                        [1, 0, 0, -1, 0, -2, 0, -1, 1, 0, 0, -1],
                        [-1, 0, 0, 1, -1, 2.5, -3.5, 0, 2 ,-1 ,1, 0],
                       [0, 1,-0.45, 0,-2.78, 0, -3.4, 0, -1, 1, -1, 0.25],
                        [0, -1, 1, 0, 1, 0, 1, 3, -2, 0,1,0]])
constr1_rhs_b=np.array([88.5, 35.4, 47.9, 41.43, 22.7])

In [8]:
#The following numpy arrays will be useful for setting the lower bound and upper bounds for the variables
lower_bound = np.array([0, 0, -np.inf, -np.inf, -34, 0, -67, 0, -23, -5, 6, 1])
upper_bound = np.array([12, 503, 4, np.inf, np.inf, 2, 20, 55, 23, 15, np.inf, np.inf])

In [9]:
# set of column indices: since we have N variables, we can take column indices to be the list [0,1,2,...,N-1]
col_indices = np.arange(N)
print(col_indices)



[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [10]:
row_indices=np.arange(M)

In [11]:
#declare the decision variables in the model
model.x = Var(col_indices)

In [12]:
# create a ConstraintList to hold multiple constraints
model.constraints = ConstraintList()

In [13]:
for i in row_indices:
   model.constraints.add(sum(constr1_coef_A[i][j]*model.x[j] for j in col_indices) <= constr1_rhs_b[i])

In [14]:
#add the upper bound and lower bounds for the variables
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [15]:
# add the model objective
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [16]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

1 Var Declarations
    x : Size=12, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :   0.0 :  None :  12.0 : False :  True :  Reals
          1 :   0.0 :  None : 503.0 : False :  True :  Reals
          2 :  None :  None :   4.0 : False :  True :  Reals
          3 :  None :  None :  None : False :  True :  Reals
          4 : -34.0 :  None :  None : False :  True :  Reals
          5 :   0.0 :  None :   2.0 : False :  True :  Reals
          6 : -67.0 :  None :  20.0 : False :  True :  Reals
          7 :   0.0 :  None :  55.0 : False :  True :  Reals
          8 : -23.0 :  None :  23.0 : False 

In [17]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155680 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [18]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1951.831
  Upper bound: -1951.831
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 13
  Number of nonzeros: 38
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01152491569519043
# ----------------------------------------------------------
#   Solution Information
# ---------------------------

In [19]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  -1951.8309999999997

Decision Variables
x[ 0 ] =  12.0
x[ 1 ] =  0.0
x[ 2 ] =  4.0
x[ 3 ] =  119.82
x[ 4 ] =  -34.0
x[ 5 ] =  0.0
x[ 6 ] =  20.0
x[ 7 ] =  0.0
x[ 8 ] =  -7.46000000000002
x[ 9 ] =  15.0
x[ 10 ] =  6.0
x[ 11 ] =  1.0

Constraints
constraints : Size=5
    Key : Lower : Body                : Upper
      1 :  None :               -47.0 :  88.5
      2 :  None : -116.28000000000002 :  35.4
      3 :  None :   47.89999999999995 :  47.9
      4 :  None :   41.43000000000002 : 41.43
      5 :  None :   10.92000000000004 :  22.7


the optimal cost are = -1951.8309999999997
and the dicision variable are 
x 0  =  12.0
x 1  =  0.0
x 2  =  4.0
x3  =  119.82
x 4  =  -34.0
x 5  =  0.0
x 6  =  20.0
x7  =  0.0
x 8  =  -7.46000000000002
x 9  =  15.0
x10  =  6.0
x11  =  1.0
the constriant1 are inactive,
the constriant1 are inactive,
the constriant1 are active,
the constriant1 are active,
the constriant1 are inactive,



---

